### PR TITLE
Add Edgecontainer api TF support

### PR DESCRIPTION
**Pull request by rayccz2**
The PR is for adding new resource for google edgecontainer API. The resources include cluster, node-pool, and vpn-connection.

If this PR is for Terraform, I acknowledge that I have:

- [ X] Searched through the [issue tracker](https://github.com/hashicorp/terraform-provider-google/issues) for an open issue that this either resolves or contributes to, commented on it to claim it, and written "fixes {url}" or "part of {url}" in this PR description. If there were no relevant open issues, I opened one and commented that I would like to work on it (not necessary for very small changes).
- [ X] Ensured that all new fields I added that can be set by a user appear in at least one [example](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/templates/terraform/examples) (for generated resources) or [third_party test](https://github.com/GoogleCloudPlatform/magic-modules/tree/main/mmv1/third_party/terraform/tests) (for handwritten resources or update tests).
- [ X] [Generated Terraform providers](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/docs/getting-started/generate-providers.md), and ran [`make test` and `make lint`](https://googlecloudplatform.github.io/magic-modules/docs/getting-started/run-provider-tests/#run-unit-tests) in the generated providers to ensure it passes unit and linter tests.
- [ X] [Ran](https://github.com/GoogleCloudPlatform/magic-modules/blob/main/docs/content/develop/run-tests.md) relevant acceptance tests using my own Google Cloud project and credentials (If the acceptance tests do not yet pass or you are unable to run them, please let your reviewer know).
- [ X] Read [Write release notes](https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/) before writing my release note below.

**Release Note Template for Downstream PRs (will be copied)**

```release-note:new-resource
`google_edgecontainer_cluster`
`google_edgecontainer_node_pool`
`google_edgecontainer_vpn_connection`
```

### CODE REVIEW

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [inline mmv1/products/edgecontainer/Cluster.yaml:25]: I think this can be removed because this value is the default

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [inline mmv1/products/edgecontainer/Cluster.yaml:97]: Per our conversation, this should probably use a diff suppress to allow server-owned labels as well as user-owned labels. See mmv1/templates/terraform/constants/cloud_run_domain_mapping.go.erb for an example.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [inline mmv1/products/edgecontainer/Cluster.yaml:109]: Can `project` be updated in-place? If not, then when users try to change the value in Terraform, we need to recreate the resource in order to get the infrastructure to match their config. You can do this by adding `immutable: true`.

Would you mind applying this rule to the other fields on these resources? From a quick scan, it appears there could be others that need to have `immutable: true` (the cidr blocks jumped out due to their description).

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [inline mmv1/products/edgecontainer/Cluster.yaml:222]: I think some of these nested fields should be `required: true` in the context of their parent, unless there is meaning when they are not specified but the parent is, ie. `maintenance_policy: {}`

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [inline mmv1/products/edgecontainer/Cluster.yaml:393]: Could `status` just be a string, since it is output-only?

File: mmv1/products/edgecontainer/NodePool.yaml
Comment (roaks3) [inline mmv1/products/edgecontainer/NodePool.yaml:65]: Should these first 3 be listed as `parameters`?

File: mmv1/products/edgecontainer/product.yaml
Comment (roaks3) [inline mmv1/products/edgecontainer/product.yaml:1]: Nit: 2023

**Review by roaks3 — CHANGES_REQUESTED** [review]
I did a first pass and had some initial comments on the fields specifically. I think many of my comments apply in multiple places, so it will help if you could try to get everything corrected at once (it will save review passes).

One other note: it looks like you were able to get tests working for the clusters (at least for simple cases), so we should probably add an [update test](https://googlecloudplatform.github.io/magic-modules/develop/test/#add-an-update-test) to ensure field behavior (you should be able to use the [generated test from this PR](https://github.com/modular-magician/terraform-provider-google-beta/compare/auto-pr-8777-old..auto-pr-8777#diff-5a3106ba15e3b7dd7ee0ac27b4d5eff0714b040314754ffec39eb29c5d3db3be) in that guide).

**Comment by roaks3** [comment]
I will be on PTO in the near-term, so re-rolling the reviewer. I can't clear my review at this point, but if it ends up blocking the merge of this change, someone should be able to override it.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:109]: I followed the edgecontainer one platform API proto setup, and only mark the field immutable: true if it has been setup the same way in the API proto.
I added the immutable: true to the fields that can not be updated in the TF resource update.

File: mmv1/products/edgecontainer/product.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/product.yaml:1]: Done.

File: mmv1/products/edgecontainer/NodePool.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/NodePool.yaml:65]: Done.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:222]: The maintenancePolicy is an optional field, but if the maintenancePolicy is provided the window need to provided. 

But I am wondering will the `required: true` on the children filed got ignored if the parent filed is unset?

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:393]: I am following the initial setup in the API proto, where the status is an enum https://source.corp.google.com/piper///depot/google3/google/cloud/edgecontainer/v1main/resources.proto;rcl=564533502;l=279

I think we might need to follow the same approach as the output from the API is enum value.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:28]: Done.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:25]: Done.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:97]: The backend logic will return the system labels + user defined labels, so I think mark it as default_from_api: true will always take as the source of truth.

**Comment by rayccz2** [comment]
> Tests failed during RECORDING mode: `TestAccEdgecontainerNodePool_edgecontainerLocalControlPlaneNodePoolInternalExample`[[Error message](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8777/artifacts/15f9ac45-8307-4055-9561-aa8ee3deccbc/build-log/recording_build/TestAccEdgecontainerNodePool_edgecontainerLocalControlPlaneNodePoolInternalExample_recording_test.log)] [[Debug log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8777/artifacts/15f9ac45-8307-4055-9561-aa8ee3deccbc/recording/TestAccEdgecontainerNodePool_edgecontainerLocalControlPlaneNodePoolInternalExample.log)]
> 
> Please fix these to complete your PR. View the [build log](https://storage.cloud.google.com/ci-vcr-logs/beta/refs/heads/auto-pr-8777/artifacts/15f9ac45-8307-4055-9561-aa8ee3deccbc/build-log/recording_test.log) or the [debug log](https://console.cloud.google.com/storage/browser/ci-vcr-logs/beta/refs/heads/auto-pr-8777/artifacts/15f9ac45-8307-4055-9561-aa8ee3deccbc/recording) for each test

Test failure due to TF SA permission error. Test passed locally 
```
2023/09/22 02:04:09 [DEBUG] Retry Transport: Returning after 1 attempts
2023-09-22T02:04:09.939Z [DEBUG] sdk.helper_resource: Called TestCase CheckDestroy: test_step_number=2 test_terraform_path=/usr/local/bin/terraform test_working_directory=/tmp/plugintest1763666537/work3230272962 test_name=TestAccEdgecontainerNodePool_edgecontainerLocalControlPlaneNodePoolInternalExample
2023-09-22T02:04:09.939Z [DEBUG] sdk.helper_resource: Finished TestCase: test_name=TestAccEdgecontainerNodePool_edgecontainerLocalControlPlaneNodePoolInternalExample test_step_number=2 test_terraform_path=/usr/local/bin/terraform test_working_directory=/tmp/plugintest1763666537/work3230272962
--- PASS: TestAccEdgecontainerNodePool_edgecontainerLocalControlPlaneNodePoolInternalExample (3897.77s)
PASS
ok  	github.com/hashicorp/terraform-provider-google-beta/google-beta/services/edgecontainer	3897.871s
```

**Review by hao-nan-li — COMMENTED** [review]
Could you update the changelog based on https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/?

Also for the skipped_test, could you either add a comment to skip_test in the yaml file or as a comment in the PR?

Awaiting on @roaks3 for final approval.

**Comment by rayccz2** [comment]
```release-note:new-resource
edgecontainer cluster/nodepool/vpnconnection
```

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [inline mmv1/products/edgecontainer/Cluster.yaml:77]: Skip the local control plane cluster test as we only have limited machine resources.
Instead the local control plane cluster test will be tested in the node pool test.

File: mmv1/products/edgecontainer/NodePool.yaml
Comment (rayccz2) [inline mmv1/products/edgecontainer/NodePool.yaml:51]: Skip the default node-pool test as we only have limited machine resources, and the node-pool creation is tested in edgecontainer_local_control_plane_node_pool_internal to avoid exposing the node location and machine name to public.

File: mmv1/products/edgecontainer/VpnConnection.yaml
Comment (rayccz2) [inline mmv1/products/edgecontainer/VpnConnection.yaml:49]: Skip the default node-pool test as we only have limited machine resources, and the vpn connection is not a critical feature.

**Comment by rayccz2** [comment]
> Could you update the changelog based on https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/?
> 
> Also for the skipped_test, could you either add a comment to skip_test in the yaml file or in a buganizer ticket?
> 
> Awaiting on @roaks3 for final approval.

**Comment by rayccz2** [comment]
> Could you update the changelog based on https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/?
> 
> Also for the skipped_test, could you either add a comment to skip_test in the yaml file or in a buganizer ticket?
> 
> Awaiting on @roaks3 for final approval.

Leaved comment on the skipped test.

**Comment by hao-nan-li** [comment]
> > Could you update the changelog based on https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/?
> > Also for the skipped_test, could you either add a comment to skip_test in the yaml file or in a buganizer ticket?
> > Awaiting on @roaks3 for final approval.
> 
> Leaved comment on the skipped test.



> > Could you update the changelog based on https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/?
> > Also for the skipped_test, could you either add a comment to skip_test in the yaml file or in a buganizer ticket?
> > Awaiting on @roaks3 for final approval.
> 
> Leaved comment on the skipped test.

Thanks for the comments. For the changelogs, could you put the full resource name? For example `google_edgecontainer_node_pool`

**Comment by rayccz2** [comment]
> > > Could you update the changelog based on https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/?
> > > Also for the skipped_test, could you either add a comment to skip_test in the yaml file or in a buganizer ticket?
> > > Awaiting on @roaks3 for final approval.
> > 
> > 
> > Leaved comment on the skipped test.
> 
> > > Could you update the changelog based on https://googlecloudplatform.github.io/magic-modules/contribute/release-notes/?
> > > Also for the skipped_test, could you either add a comment to skip_test in the yaml file or in a buganizer ticket?
> > > Awaiting on @roaks3 for final approval.
> > 
> > 
> > Leaved comment on the skipped test.
> 
> Thanks for the comments. For the changelogs, could you put the full resource name? For example `google_edgecontainer_node_pool`

Updated the release node with `google_edgecontainer_cluster, google_edgecontainer_node_pool, google_edgecontainer_vpn_connection`

**Review by hao-nan-li — APPROVED** [review]
LGTM, will let @roaks3 making the final approval.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [reply mmv1/products/edgecontainer/Cluster.yaml:77]: Would you mind making these comments in the yaml file directly above this line? That will make it easier for us to keep track of

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [reply mmv1/products/edgecontainer/Cluster.yaml:93]: That will create a problem though, because Terraform will try to overwrite the entire set of labels to only include what is defined by the user. The custom logic would tell Terraform to ignore certain values that are controlled by the server.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [inline mmv1/products/edgecontainer/Cluster.yaml:116]: I think this can just be changed to `output: true`, based on the API docs. That way a user can't try to set it.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [inline mmv1/products/edgecontainer/Cluster.yaml:169]: I don't see this field in the API docs, but it seems like something that should also be `output: true`. Can this be set to contradict the CIDR block configs?

Also, you probably don't need `NETWORK_TYPE_UNSPECIFIED` as an option, and if this is `output: true`, it can probably be changed to a string.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [inline mmv1/products/edgecontainer/Cluster.yaml:180]: I'm not sure what the intention is for this field, but I believe that if you want it to be provided by the server when it isn't set, you will need to add `default_from_api: true` to each nested field. If that's not what we want, I believe this line should be removed.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [reply mmv1/products/edgecontainer/Cluster.yaml:227]: Yes, the required attribute applies only within the parent

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [inline mmv1/products/edgecontainer/Cluster.yaml:289]: Here and under `local`, I think you need `default_from_api: true` if it can be populated by the server.

**Review by roaks3 — CHANGES_REQUESTED** [review]
This will probably need another pass due to the large number of fields being added, but I reviewed Cluster with some suggestions. I'm also fairly sure the `labels` field will need additional logic as mentioned in my comment.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:93]: Thanks for the explanation. Added `diff_suppress_func: 'DomainMappingLabelDiffSuppress'` in cluster/node-pool label fields.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:227]: Thanks, added required: true into the children fields of maintenancePolicy.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:415]: Gotcha, thanks for the explanation. Change the Enum -> String type for the fields that is marked as output: true.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:77]: Done.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:116]: Thanks for capturing, added output: true and remove the immutable

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:169]: Thanks for capturing, this field is output only. It is not in the public doc because currently we mark the field as hidden. 
Also change the field type from Enum -> String type.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:180]: Gotcha, adding the `default_from_api: true` to each nested field. 
The field `authorization` is to define the user credential for accessing the cluster and is required for create operation.
The server will not modify the content, so setting default_from_api to each field should be the correct approach.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:289]: Done.

**Comment by rayccz2** [comment]
> This will probably need another pass due to the large number of fields being added, but I reviewed Cluster with some suggestions. I'm also fairly sure the `labels` field will need additional logic as mentioned in my comment.

Sorry for the confusion I check the code of the labels field for cluster and node-pools it turns out we only serving the user created labels, the system created labels directly applied to the cluster and not served in the edgecontainer api service.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (roaks3) [reply mmv1/products/edgecontainer/Cluster.yaml:95]: Ah ok, makes sense, thanks!

**Review by roaks3 — CHANGES_REQUESTED** [review]
I reviewed the whole PR on this pass and things looked good. I had a few more comments, but this should be ready to merge after they are addressed.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:117]: Done.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:170]: Done.

File: mmv1/products/edgecontainer/Cluster.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/Cluster.yaml:176]: Gotcha, I thought they are not conflicting.

File: mmv1/products/edgecontainer/VpnConnection.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/VpnConnection.yaml:123]: There is no update vpnconnection api available so I assume once the vpn connection got created, all the fields are immutable.

File: mmv1/templates/terraform/examples/edgecontainer_cluster.tf.erb
Comment (rayccz2) [reply mmv1/templates/terraform/examples/edgecontainer_cluster.tf.erb:3]: Thanks for pointing out, removing the extra spaces.

File: mmv1/products/edgecontainer/VpnConnection.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/VpnConnection.yaml:81]: There is no update vpnconnection api available so I assume once the vpn connection got created, all the fields are immutable.

File: mmv1/products/edgecontainer/NodePool.yaml
Comment (rayccz2) [reply mmv1/products/edgecontainer/NodePool.yaml:110]: I go though all 3 resources and make sured the validity of the default_from_api

Comment (rayccz2) [reply mmv1/products/edgecontainer/NodePool.yaml:110]: I go though all 3 resources and make sured the validity of the default_from_api

**Review by roaks3 — APPROVED** [review]
LGTM, we can merge once main is unlocked

**Pull request merged**
